ipa-core/src/protocol/ipa_prf/oprf_padding/mod.rs

```diff
@@ -1,5 +1,866 @@
+        ctx.narrow(&PaddingDpStep::PaddingDpPass2),
+        input,
+        Role::H3,
+        Role::H1,
+        Role::H2,
+        &padding_params,
+    )
+    .await?;
+
+    // H2 and H3 add padding noise
+    input = apply_dp_padding_pass::<C, T, B>(
+        ctx.narrow(&PaddingDpStep::PaddingDpPass3),
+        input,
+        Role::H2,
+        Role::H3,
+        Role::H1,
+        &padding_params,
+    )
+    .await?;
+
+    let after_padding_len = input.len();
+    tracing::info!(
+        "Total number of padding records added: {}. Padding Parameters: {:?}",
+        after_padding_len - initial_len,
+        padding_params
+    );
+
+    Ok(input)
+}
+
+/// Apply dp padding with one pair of helpers generating the noise
+/// Steps
+///     1.  Helpers `h_i` and `h_i_plus_one` will get the same rng from PRSS
+///         and use it to sample the same random noise for padding from `OPRFPaddingDp`.
+///         They will generate secret shares of these fake rows.
+///     2.  `h_i` and `h_i_plus_one` will send the send `total_number_of_fake_rows` to `h_out`
+///     3.  `h_out` will generate secret shares of zero for as many rows as the `total_number_of_fake_rows`
+///
+/// # Errors
+/// Will propogate errors from `OPRFPaddingDp`
+/// # Panics
+/// Will panic if called with Roles which are not all unique
+pub async fn apply_dp_padding_pass<C, T, const B: usize>(
+    ctx: C,
+    mut input: Vec<T>,
+    h_i: Role,
+    h_i_plus_one: Role,
+    h_out: Role,
+    padding_params: &PaddingParameters,
+) -> Result<Vec<T>, Error>
+where
+    C: Context,
+    T: Paddable,
+{
+    // assert roles are all unique
+    assert!(h_i != h_i_plus_one);
+    assert!(h_i != h_out);
+    assert!(h_out != h_i_plus_one);
+
+    let mut total_number_of_fake_rows = 0;
+    let mut padding_input_rows: Vec<T> = Vec::new();
+
+    // Step 1: Helpers `h_i` and `h_i_plus_one` will get the same rng from PRSS
+    // and use it to sample the same random noise for padding from OPRFPaddingDp.
+    // They will generate secret shares of these fake rows.
+    if ctx.role() != h_out {
+        total_number_of_fake_rows = two_parties_add_dummies::<C, T, B>(
+            &ctx,
+            &mut padding_input_rows,
+            h_i,
+            h_i_plus_one,
+            padding_params,
+        )?;
+    }
+
+    // Step 2: h_i and h_i_plus_one will send the send total_number_of_fake_rows to h_out. h_out will
+    // check that both h_i and h_i_plus_one have sent the same value to prevent any malicious behavior
+    // See oprf_padding/README.md for explanation of why revealing the total number of fake rows is okay.
+    let send_ctx = ctx
+        .narrow(&SendTotalRows::SendNumFakeRecords)
+        .set_total_records(TotalRecords::ONE);
+    match ctx.role() {
+        role if role == h_i || role == h_i_plus_one => {
+            let direction = if role == h_i {
+                Direction::Left
+            } else {
+                Direction::Right
+            };
```

Comment:
I'd suggest adding the following helper to `impl Role`:

```rust
    /// Returns the direction to the peer with the specified role.
    ///
    /// If `self == role`, returns `None`.
    #[must_use]
    pub const fn direction_to(&self, role: Role) -> Option<Direction> {
        match (self, role) {
            (H1, H2) | (H2, H3) | (H3, H1) => Some(Direction::Right),
            (H1, H3) | (H2, H1) | (H3, H2) => Some(Direction::Left),
            (H1, H1) | (H2, H2) | (H3, H3) => None,
        }
    }
```

Then, you can:
1. Replace lines 401-407 with:
```rust
    if let Some(direction) = ctx.role().direction_to(h_out)
```
2. Change the `_h_out` branch to an `else`, like we talked about this morning.
3. Change `two_parties_add_dummies` and `add_padding_items` to take a `Direction` rather than `Role`s.

It may also make sense to move steps 1 and 3 into the if/else cases.

Reply:
okay, that's nice to simplify some of the role checking in `apply_dp_padding_pass`.   I'm not sure I follow how to only use a `Direction` in `add_padding_item` as will need to compare with `ctx.role()`

Reply:
What I was thinking is replace the tests against `ctx.role()` with:

In `two_parties_add_dummies`:
```rust
// If the excluded helper is on our left, we want to use the PRSS shared with the
// helper on our right, and vice-versa.
let (mut left, mut right) = ctx.prss_rng();
let rng = match direction_to_excluded_helper {
    Direction::Left => &mut right,
    Direction::Right => &mut left,
};
```

In `add_padding_items` for `OPRFIPAInputRow`:
```rust
let match_key_shares = match direction_to_excluded_helper {
    Direction::Left => AdditiveShare::new(BA64::ZERO, dummy_mk),
    Direction::Right => AdditiveShare::new(dummy_mk, BA64::ZERO),
};
```

... and similarly in the other `add_padding_items` implementation. This initializes these variables without requiring mutability.

Reply:
Okay, got it.   Following in that direction was able to simplify elsewhere as well to only pass in the `excluded_helper: Role` to `apply_dp_padding_pass`.